Improve Presto Verifier

caithagoras opened the pull request:
Improve the existing verifier.

### Behavior differences
- Fix the issue where result-mismatched insert queries are not reran for non-determinism check.
- Fix the issue where checksum queries are treated as main queries while main queries are treated as part of the setup queries, leading to various mis-information on the UIs.
- Select is rewritten into `CreateTableAsSelect` and checksum queries are used for result comparison. This removes the row count limit for Select queries verification.
- Floating point columns are compared using relative errors.
- Orderable array columns are applied with `array_sort` before `checksum`.
- Verification happens along with rewrite, to avoid the super-long wait for the rewrite to complete before verification even starts.
- Query timeout are now enforced by session properties instead of by local timer.
- Add retries for query failures due to network errors and presto retryable errors.
- Add configuration properties `metadataTimeout` and `checksumTimeout`, the timeout duration for different Presto queries.
- Add configuration property `human-readable.log-file` to allow human readable events to be logged to file. Streamline human readable event logging.
- Add integration smoke test.
- Remove a bunch of unnecessary configuration properties.
- Make verifier easier to configure for community users.

### Changes to output event
  - Add `EventStatus status`, to indicate whether is verification is succeeded, failed, skipped or failed but auto-resolved.
  - Add `deterministic` to indicate whether result mismatch is due to non-determinism.
  - Add `originalControlQuery` and `originalTestQuery`, the control and test queries before the rewrite.
  - Add `controlChecksumQuery` and `testChecksumQuery`.
  - Add `errorCode`.
  - Add `mismatchedColumns`, a mapping from mismatched column names to their types.
  - Remove `source`.

caithagoras commented:
@wenleix @nezihyigitbasi @rongrong @mayankgarg1990 @shishunzhong @mbasmanova 
Thank you all for you inputs regarding how we should roll out the new verifier.

Here are my thoughts:
- We should engage with the community to get feedback and feature requests about the verifier and related tooling in order for anyone to run large-scale verification. I have since created a Slack channel `#releaseverification` for discussion about these topics. Making the verifier rewrite open-source allows contributors to see the current status of the new verifier so that they can provide meaningful feedback and improvement suggestions.
- Since we mostly agree that there are very few usage of the existing verifier outside of Facebook, it would be the easiest to delete the existing code and replace it with the new version, instead of having a separate module named as `presto-verifier-v2`. The old verifier will always be accessible, if needed, through previous releases.
- We should have the same standard of code quality of both open source and Facebook-specific repo, and thus I'm rewriting the verifier have that in mind. There will certainly be community demands beyond this PR and we can also make improvements to it.

rongrong commented:
So the new verifier will not have the same API as the old one? If we don't have to change how you launch verifier, I'd restructure the changes as: diffs introducing new features / code, diff to switch the launcher to use new code and remove inaccessible code. Deleting the existing verifier before any replacement is introduced looks weird from commits perspective.

caithagoras commented:
@rongrong Nezih mentioned this approach as well. It turns out not feasible to have both launcher in the same maven module while allowing both launchers to be accessible. To keep both at the same time, we inevitably have to create a separate sub-module `presto-verifier-v2`.

caithagoras commented:
@rongrong I was not clear why I would prefer to remove the old version.

Although we're able to keep old and new version within the same module using 1 launcher + different flags. I see some minor downsides and benefits, none of which make it a strong argument to go either way, except for maybe operational easiness.
- Downside: pom.xml will contains a mixture of dependencies.
- Downside: There will be classes with the same name but different content, though the fix would be to move the existing classes into a legacy sub-package.
- Upside: We can make sure that v2 is solid before deleting the old code, though external and internal users will be able access/launcher either version even if the old code is deleted upfront.

rongrong commented:
If would be nice to see whether anyone in the community is using verifier and what's their take on this. If the deletion of old version and addition of new version happens in the same PR (same release) there's not really any practical issue. So I won't hold strong opinions on how the commits should be structured within a single PR.

mayankgarg1990 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ChecksumResult.java:
```diff
@@ -0,0 +1,91 @@
+                    checksum == null ? "null" : checksum.getClass().getName());
+        }
+        this.rowCount = rowCount;
+        this.checksums = unmodifiableMap(requireNonNull(checksums, "checksums is null"));
```
`requireNonNull(checksums)` should go to the top because we would have already accessed it in `checksum.values()`

mayankgarg1990 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ColumnValidator.java:
```diff
@@ -0,0 +1,32 @@
+
+abstract class ColumnValidator
+{
+    abstract List<SingleColumn> generateChecksumColumns(Column column);
```
nit `generateValidationColumns`. Not all the validations are checksums.

caithagoras replied:
Address here: https://github.com/prestodb/presto/pull/12489#issuecomment-475096230

mayankgarg1990 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ChecksumResult.java:
```diff
@@ -0,0 +1,91 @@
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+
+public class ChecksumResult
```
We aren't just running checksums. How about naming this as `QueryValidationOutput`. Since in this case, we executed a validation query (which consists of checksums and more)

caithagoras replied:
Same as above, addressed here: https://github.com/prestodb/presto/pull/12489#issuecomment-475096230

mayankgarg1990 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ChecksumResult.java:
```diff
@@ -0,0 +1,91 @@
+public class ChecksumResult
+{
+    private final long rowCount;
+    private final Map<String, Object> checksums;
```
same as above - `outputs`?

caithagoras replied:
Same as above: https://github.com/prestodb/presto/pull/12489#issuecomment-475096230

mayankgarg1990 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ChecksumValidator.java:
```diff
@@ -0,0 +1,71 @@
+import static com.facebook.presto.verifier.framework.Column.Category.SIMPLE;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+public class ChecksumValidator
```
Going with my comment above - how about `OutputValidator`

caithagoras replied:
Same as above: https://github.com/prestodb/presto/pull/12489#issuecomment-475096230

rongrong replied:
I agree with @mayankgarg1990 here that since there's a checksum function, which is heavily used in the validation query, it's a bit misleading. The `ChecksumValidator` here obviously include non-checksum-function validations. I'm actually ok with just calling this `Validator`. `Output` seems to be redundant.

caithagoras replied:
There are multiple occurrences of the word `checksum` both in code and in general. For example, we always say `checksumQuery`, and everyone know what that means, but if we use `validationQuery`, it is not clear whether it is referring to the query that produces the comparison datum, or to the query being validated itself.

mayankgarg1990 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/FloatingPointColumnValidator.java:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    @Override
+    boolean isMatched(Column column, ChecksumResult controlResult, ChecksumResult testResult)
```
Do you feel it will be useful to return the result of validation in a better way? Roughly, the following is what I am thinking of:

```
class ColumnMatchResult {
  boolean isMatch;
  Optional<String> errorMessage;
}
```

The idea being to be able to return custom exception messages. Example - `The sums of test (1234.43545) and control (1345.34353) were off by x%` So that the criteria for test failure is clear.

caithagoras replied:
This will be useful, but I would delay it to next iteration, as it make the output more complicated. We need to think about how we're going to represent them in the output event, as well as, how we're going to store and present the data to the UI

mayankgarg1990 replied:
We already have a `errorMessage` field - so we can just put it as text there. We can have that in the first iteration.

caithagoras replied:
The `errorMessage` field is parsed by some php code to present for the UI at this time, so we probably should not break its contract here.

mayankgarg1990 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/retry/RetryDriver.java:
```diff
@@ -0,0 +1,89 @@
+                attempt++;
+                int delayMillis = (int) min(minBackoffDelay.toMillis() * pow(scaleFactor, attempt - 1), maxBackoffDelay.toMillis());
+                int jitterMillis = ThreadLocalRandom.current().nextInt(max(1, (int) (delayMillis * 0.1)));
+                log.debug(
```
I believe we should not print the full stack trace - but print the fact that we are retrying a query.

caithagoras replied:
No stacktrace is printed for readability reason, only `callableName` and `qe.getErrorCode` is printed. Also, this is a debug-level logging.

mayankgarg1990 replied:
I wasn't clear here. 

I meant that we should have a `log.info` and actually put this in logs. And since we are not printing the stack trace - it should be fine. So a convoluted way to say - lets convert this to `log.info`

caithagoras replied:
Making it info-level logging will pollute the output. It's not very helpful for anyone who read the log to see a bunch of retry logging. Since we're going to log all failures in the output. We should build a dashboard so that release oncall read its signal easily.

mayankgarg1990 replied:
As I look into the existing logs from the current verifier (assuming that its logging is similar to the new one - if not please call that out), this is what we print in the logs:

```
2019-03-12T02:45:43.636-0700	INFO	main	stdout	Name: 20190305_220551_29858_fas8x

2019-03-12T02:45:43.636-0700	INFO	main	stdout	Schema (control): bizapps

2019-03-12T02:45:43.636-0700	INFO	main	stdout	Schema (test): bizapps

2019-03-12T02:45:43.636-0700	INFO	main	stdout	Valid: true

2019-03-12T02:45:43.636-0700	INFO	main	stdout	Query (test): SELECT "PC_hdloaner_onhand_02"."binsize" AS "binsize",

  "PC_hdloaner_onhand_02"."locator_code" AS "locator_code",

  "PC_hdloaner_onhand_02"."onhand_qty" AS "onhand_qty",

  "PC_hdloaner_onhand_02"."operating_system" AS "operating_system",

  "PC_hdloaner_onhand_02"."reorderpoint" AS "reorderpoint"

FROM (

  select * from hdloaner_onhand_02

) "PC_hdloaner_onhand_02"

2019-03-12T02:45:43.636-0700	INFO	main	stdout	Control Duration (secs): 301.826719127

2019-03-12T02:45:43.636-0700	INFO	main	stdout	   Test Duration (secs): 25.136472249999997

2019-03-12T02:45:43.636-0700	INFO	main	stdout	Control CPU (secs): 7469.8240000000005

2019-03-12T02:45:43.636-0700	INFO	main	stdout	   Test CPU (secs): 15275.392

2019-03-12T02:45:43.636-0700	INFO	main	stdout	         CPU Ratio: 2.0
```
per query.

So, if we retry twice and print 2 more lines, I am not sure how that pollutes the logs. Similar to how we are printing in the logs that we started the query, this is the result and the CPU comparison, we should also print failures.

caithagoras replied:
This is what I wanted to change - the new version will not output any of those logs. We run 10000 queries per suite, it would be impossible for human to process those log. Instead, we should only output information about the overall status and progress of the verifier. Each individual query event should be queried through xdb or presented through UI.

mayankgarg1990 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/framework/PrestoAction.java:
```diff
@@ -0,0 +1,247 @@
+            Optional<ResultSetConverter<R>> converter)
+    {
+        return prestoRetry.run(
+                "presto",
```
Should there be a more meaningful name passed here? I am thinking about the following options (multiple can be included) - 

1) kind of retry - network or presto
2) some query specific identifier
3) stage of the query - setup, main, non-deterministic, teardown

caithagoras replied:
Add 1 and 3, not sure about what do you mean by "some query specific identifier"
I'm not too worry about this since it is for debugging logging only.

mayankgarg1990 replied:
> some query specific identifier

I meant something which identifies the specific query that we are testing so that if we see this error message, we can trace it back to a specific query. I did have a comment somewhere which talked about actually logging about the failures. These can be helpful during debugging. So something which we can do if we are trying to figure out what happened to a specific test case.

mayankgarg1990 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/retry/RetryDriver.java:
```diff
@@ -0,0 +1,89 @@
+            }
+            catch (QueryException qe) {
+                if (attempt >= maxAttempts || !retryPredicate.test(qe)) {
+                    throw qe;
```
In my opinion, we need a way to record failed runs. Specifically - if some changes increase the inter-worker communication failures, we want to record that information - and surface it up. So, I want to able to see on the UI, queries are mostly succeeding in their 3rd try as compared to in the first one. Also, which errors happened will be useful information to have.

caithagoras commented:
@mayankgarg1990 Thanks for reviewing!
Regarding your comments about checksum vs. validation. I'm thinking `checksum` more of as the word in English in general, but not referring to the Presto function `checksum`, or the hashing algorithm. According to Wikipedia:
```
A checksum is a small-sized datum derived from a block of digital data for the purpose of detecting errors that may have been introduced during its transmission or storage.
```
And `ChecksumResult` and `generateChecksumColumns` seems to fit this definition as they are small-size data for verification purpose.

The reason I was avoiding "validation" is that:
- `validation` more about the process of comparing and finding the difference between the control result and the test result, while `ChecksumResult` and `generateChecksumColumns` are data derived from a single query, either control or test.
- `validation` has similar meaning as `verification`, so I would stick to having only one (verification) of the two words appearing in this project.

caithagoras commented on presto-verifier/src/main/java/com/facebook/presto/verifier/event/VerifierQueryEvent.java:
```diff
@@ -0,0 +1,312 @@
+    private final String name;
+
+    private final boolean succeeded;
+    private final boolean skipped;
```
I would slightly prefer the way it is. Enum may make the Java code look nice, but they are eventually converted to string in the json output, and its `one-of` property gets lost. Clients see a string field and needs to process it based on some hard-coded values, whereas if boolean is used, clients only need to know the field name, since its value is binary.

mayankgarg1990 replied:
At the same time that leads us to think of situations, can both `succeeded` and `skipped` be true and then check the different fields. Having a single value that states clearly the situation is better - you just need to read a single field and then these values are familiar enough for someone to infer.

caithagoras commented on presto-verifier/src/main/java/com/facebook/presto/verifier/framework/PrestoExceptionClassifier.java:
```diff
@@ -0,0 +1,116 @@
+            NO_NODES_AVAILABLE,
+            REMOTE_TASK_ERROR,
+            SERVER_SHUTTING_DOWN,
+            SERVER_STARTING_UP,
```
We shouldn't retry `GENERIC_INTERNAL_ERROR`, as the nature of the failure is unknown.
For the other 4 failures, I believe they reliability related (correct me if I'm wrong), so I would like to add them but wait for the commit that monitors query failures.

mayankgarg1990 replied:
Can we add a `TODO` which talks about reconsidering this list and updating when we add retry tracking

caithagoras replied:
This is now added to the last commit

caithagoras replied:
Since all failures occurred during a verification is exported, we'll be able to tell the difference that you mentioned.

caithagoras replied:
I have already added those 4 failures. See commit `Export all query failures for each verification `

caithagoras replied:
I have already added those 4 failures. See commit "Export all query failures for each verification"

caithagoras commented on presto-verifier/src/main/java/com/facebook/presto/verifier/framework/PrestoExceptionClassifier.java:
```diff
@@ -0,0 +1,116 @@
+            HIVE_WRITER_CLOSE_ERROR,
+            HIVE_FILESYSTEM_ERROR,
+            HIVE_CANNOT_OPEN_SPLIT,
+            HIVE_METASTORE_ERROR);
```
The above list is from our validator code for the hql-translator project and they've being production-tested. I think we hold-off adding more until we see cases that are in those category could succeed with retry.
For example, `HIVE_WRITE_VALIDATION_FAILED` does sounds to be retryable to me by the look of the name.

mayankgarg1990 replied:
The only downside here is that we retry the queries. `HIVE_WRITE_VALIDATION_FAILED` as you pointed out might not fit - but I feel that there are other error codes that we can and should consider. As you said, we should refer to production data and figure this out.

caithagoras commented on presto-verifier/src/main/java/com/facebook/presto/verifier/framework/QueryOrigin.java:
```diff
@@ -0,0 +1,76 @@
+        MAIN,
+        TEARDOWN,
+        DESCRIBE,
+        CHECKSUM,
```
Same as above: https://github.com/prestodb/presto/pull/12489#issuecomment-475096230

caithagoras commented on presto-verifier/src/main/java/com/facebook/presto/verifier/framework/AbstractVerification.java:
```diff
@@ -0,0 +1,298 @@
+    protected QueryConfiguration getConfiguration(QueryGroup group)
+    {
+        checkState(group == CONTROL || group == TEST, "Unexpected QueryGroup %s", group);
+        return group == CONTROL ? sourceQuery.getControlConfiguration() : sourceQuery.getTestConfiguration();
```
I'm not sure what is the standard for this case in Presto. Using a `switch` statement makes this 2 lines of code 7 lines, so it's hard for me to justify it's more readable with a `switch`.

mayankgarg1990 replied:
This is not a big thing - I am just not able to wrap my head around testing for enum values. But thats fine :) The reason why you have the checkState is to ensure that when new values are added, the code acts properly - but when you add the next enum (if it ever happens) - then the return statement gets converted to an if-else which will be 9 lines. So, I think it is fine to delay this change to when we end up writing another `group`

caithagoras commented on presto-verifier/src/main/java/com/facebook/presto/verifier/event/VerifierQueryEvent.java:
```diff
@@ -0,0 +1,312 @@
+    private final String controlChecksumQuery;
+    private final String controlQueryId;
+    private final Double controlCpuTimeSecs;
+    private final Double controlWallTimeSecs;
```
Great. I'll massage them into `QueryInfo`

caithagoras commented on presto-verifier/src/main/java/com/facebook/presto/verifier/framework/VerificationReport.java:
```diff
@@ -0,0 +1,124 @@
+
+public class VerificationReport
+{
+    public enum MatchType
```
renamed to `VerificationResult `

mayankgarg1990 commented:
>  Regarding your comments about checksum vs. validation. I'm thinking checksum more of as the word in English in general, but not referring to the Presto function checksum, or the hashing algorithm. According to Wikipedia:

I actually did exactly the same before my initial comments :). My main concern here is coming from the fact that we probably relate it too closely with the checksum function and the crypto definitions and the fact that we have a checksum function present doesn't help either. So going for something more general which doesn't have this ambiguity will be better in my opinion.

mayankgarg1990 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/framework/VerificationContext.java:
```diff
@@ -0,0 +1,38 @@
+
+public class VerificationContext
+{
+    private Map<QueryGroup, ListMultimap<QueryStage, String>> failures = new EnumMap<>(QueryGroup.class);
```
This is insufficient information in my opinion. To debug the issues we need to have more information. At the least, I am thinking about having the following:

1) presto query id (if we could launch the query)
2) full stack trace
3) error code (something you already have)

caithagoras replied:
I think the purpose of exporting each individual failures including the retried ones is to see if there are systematic reliability issue. I'm ok with adding presto query id so that release oncalls can have more power if they want to go and see individual failures (the ones that failed but retried successfully). This is actually already a step involving human effort and ideally we should minimize release oncall having to do this. Stacktrace can be accessed with the query id so I don't think it would be necessary to duplicate them in order xdb.

mayankgarg1990 replied:
> This is actually already a step involving significant human effort and ideally we should minimize the need for release oncall to do this. (i.e. we should build dashboard around this allFailures fields to show aggregated failure reasons, etc.)

An aggregated view will be good - but during debugging showing more information is better than less. If it feels noisy, it is easier and quicker to remove information from the UI as compared to pumping it into the tooling and the flowing through the DB to the UI.

> Stacktrace can be accessed with the query id so I don't think it would be necessary to duplicate them in order xdb.

As long as we can fetch this information on the UI, I am fine with whichever way you feel is better. The only case where I think this might not be possible with a query id is when we fail launching the query - in that situation as per my understanding since the query didn't even reach presto, we don't even have a query id.

caithagoras replied:
I have now finalized the last commit, please take a look again.

rongrong commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ColumnValidator.java:
```diff
@@ -0,0 +1,32 @@
+
+    abstract boolean isMatched(Column column, ChecksumResult controlResult, ChecksumResult testResult);
+
+    static Identifier identifier(String name)
```
Is this function the reason this is an abstract class rather than an interface? Does it make more sense to have a `getIdentifier` function in the `Column` class instead?

caithagoras replied:
Good catch. Yes, I'm converting this to an interface and move the static method into `VerifierUtil#delimitedIdentifier`. Also adding `Column#getIdentifier`.

rongrong commented on presto-verifier/src/main/java/com/facebook/presto/verifier/framework/QueryBundle.java:
```diff
@@ -0,0 +1,65 @@
+public class QueryBundle
+{
+    private final Optional<QualifiedName> tableName;
+    private final List<Statement> setupQueries;
```
It would be nice to have a tighter alignment between these and `QueryStage`.

caithagoras replied:
I don't see a good way and I'm not sure if that's natural thing to do, but open to suggestions.

caithagoras commented:
- Moved the event client commit, to put it after the framework commit. `VerifierQueryEvent` and `QueryInfo` are merged into the framework commit, so they are declared and used in the commit.
- Converted `boolean succeeded`, `boolean skipped`, and `boolean resolved` into `EventStatus status` in the output event.
- Massage query-related common fields in `VerifierQueryEvent` into `QueryInfo`.
- `mismatchedColumns` is no longer exported as an event field. Column match details are included in the error message.
- Added a commit export all query failures so that we'll insights about reliability issues
- Added a commit to support automatic resolution of failures

mayankgarg1990 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/framework/VerificationContext.java:
```diff
@@ -0,0 +1,54 @@
+        return failures.get(group).stream()
+                .map(exception -> new FailureInfo(
+                        exception.getQueryOrigin().getStage(),
+                        exception.getErrorCode(),
```
Looking at how `QueryException` is generated for Cluster connection issues:

```
    public static QueryException forClusterConnection(Throwable cause)
    {
        return new QueryException(cause, CLUSTER_CONNECTION, Optional.empty(), true, Optional.empty());
    }

```
the error code is empty. So in this situation, for cluster connection issues - we will essentially pass in an empty error code and there will be no information about what happened. Should we not just serialize and return the whole `QueryException` object?

caithagoras replied:
- `getErrorCode` is different from `getPrestoErrorCode` - it is always a non-empty string. It is a short description of the error. Examples are `PRESTO(PERMISSION_ERROR)` or `CLUSTER_CONNECTION(SocketTimeoutException)`

mayankgarg1990 replied:
Thanks for the clarification on the `getErrorCode` method - I missed that.

Re: stacktrace: atleast based on my previous work in Hive - stack traces have been an integral part of debugging. I don't have an exact example that I can cite as of now to justify it, but overall, I can't imagine digging into specifics without stack traces. If you just give me `SocketTimeoutException` that gives me very few details and pointers to start debugging.

Since you have been resistant to add it, I am curious to hear your point of view of not adding stack traces. If the only reason is that you feel it is not useful, I will differ in opinion here. Is there some other reason as well?

caithagoras commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ColumnMatchResult.java:
```diff
@@ -0,0 +1,71 @@
+    private final boolean matched;
+    private final Optional<String> message;
+
+    public ColumnMatchResult(boolean matched, Optional<String> message)
```
This is not the case here. Currently only floating point columns have message. For simple columns and array columns, what can be included in the message? (I don't think it would be helpful to expose the checksum that differentiate between control and test for those columns)

caithagoras replied:
Though I can add a check such that when `matched`, `message` cannot be present.

mayankgarg1990 replied:
I was about to recommend the second, but when I looked at the code:

```
return new ColumnMatchResult(
                    controlSumObject == null && testSumObject == null,
                    Optional.of(format("control(sum: %s) test(sum: %s)", controlSumObject, testSumObject)));
```

There are situations that `matched = true` and the message is not empty. I felt that extra information doesn't hurt :). We can always decide what we want to show in the UI vs not.

> (I don't think it would be helpful to expose the checksum that differentiate between control and test for those columns)

Extra information in case of failure doesn't hurt. I am fine with just printing the checksums as well in that case. My only concern is to tighten the API so that when we add new checks, it ensures that people add a message.

caithagoras replied:
Oh, now I remembered that's why I called it `message` not `errorMessage`. I'll add checksum information, but in terms of tightening the API, this is not very feasible.
- Merely adding `checkArgument` is not limiting unwanted caller behavior, and cannot be detected until at runtime.
- To tighten the API, we need to make the constructor private and to private static factory constructor such as `matched()` and `mismatched(String)`. However, since `boolean matched` can sometimes be an expression, rather than always a constant `true` or `false` at a caller, it makes those constructors very hard to use.

mayankgarg1990 replied:
> To tighten the API, we need to make the constructor private ... 

I understand this and was my first thought before I looked at the matched message use case.

> Merely adding checkArgument is not limiting unwanted caller behavior, and cannot be detected until at runtime.

I completely agree. If someone hopefully add tests or tests in prod we should catch hold of it.

caithagoras commented on presto-verifier/src/main/java/com/facebook/presto/verifier/framework/Column.java:
```diff
@@ -0,0 +1,124 @@
+    private final Category category;
+    private final Type type;
+
+    public Column(String name, Category category, Type type)
```
It is used in tests, as it would be inconvenient to use `fromResultSet` where we simply want to create a `Column`. Added annotation `@VisibleForTesting`.

caithagoras commented:
regd: https://github.com/prestodb/presto/pull/12489#discussion_r268907447
I've already added the 4 error type to  commit "Export all query failures for each verification""

caithagoras commented:
regd logging:
- Debug logging can always be enabled from the pt command that launches the verifier, using `log.properties`. Log level is designed to control output pollution and thus has been built in airlift - we don't need a special configuration property for that purpose.
- The "logging" that you see in the existing verifier does not serves the purpose of "Being able to debug unknown behaviors". The existing logs were actually from `HumanReadableEventClient`, which dumps `VerifierQueryEvent` to stdout. When it prints, the verification for a specific query has already been completed and event has already been generated. If there is a bug causing the process or a thread to be stuck in the middle of execution, the event will simply not be emitted. That being said, the old verifier does not have much info-level logging for the purpose of "checkpointing" for each thread (other than the progress reporting, which we do have in the newer version as well).
- If we want to introduce logging to be able to see if threads are stuck, we need some more clever design - simply log.info everywhere will not help human debugging when 100+ threads are running at the same time. What we can do, for example, is to have a monitoring thread to periodically report # of threads that are working, and/or threads that have not completed any verification for a long period, and/or if those long-running verification is due to a long-running Presto query, etc.

caithagoras commented:
@mayankgarg1990 Thanks again for reviewing! Here're some post-approval changes I made:
- Added interface `SourceQuerySupplier`. In "Support supplying verification source queries through MySQL", removed awkward `getSourceQueriesSupplierProvider` and replaced it with a plugin-style source query supplying mechanism.
- Allowed xdb table name to be specified by configuration property `source-query.table-name` (might be useful for cbo testing purpose)
- Added `ExceededTimeLimitFailureResolver`.
- Updated Verifier installation guide. I'll add detailed verifier mechanism in a separate PR.
